esm: add chaining to loaders (#42623)
nodejs/node: doc/api/esm.md

When hooks are used they apply to the entry point and all `import` calls. They
won't apply to `require` calls; those still follow [CommonJS][] rules.

Loaders follow the pattern of `--require`:

```console
node \
  --experimental-loader unpkg \
  --experimental-loader http-to-https \
  --experimental-loader cache-buster
```

These are called in the following sequence: `cache-buster` calls
`http-to-https` which calls `unpkg`. Or in terms of simple functions,

> `cacheBuster(httpToHttps(unpkg(input)))`.

Author: Perhaps this is not as clear as it could be, but I think it is accurate: in loaders-world, `https-to-https::resolve()` literally is called by `cache-buster::resolve()`.

The bit about results being passed I think is less clear. Perhaps I should remove the `cacheBuster(httpToHttps(unpkg(input)))` altogether since it's not 100% how it works?

I'm not married to this wording, but I think the suggestion is not clearer.

> […] loaders are invoked in the order they are presented on the command line.

This is not true (and it's the confusing part): they are invoked in **reverse** order of how they were passed in the command-line.

> `cacheBuster(httpToHttps(unpkg(input)))`.

Reviewer: > This is not true (and it's the confusing part): they are invoked in reverse order of how they were passed in the command-line.

That's not what this example shows: `unpkg` is the first command-line argument and `unpkg(input)` is the first function to be called too.

> `cacheBuster(httpToHttps(unpkg(input)))`.

Reviewer: I think removing `cacheBuster(httpToHttps(unpkg(input)))` is the best thing to do, because it's this syntax that makes it look reversed.

### Hooks

Hooks are part of a chain, even if that chain consists of only one custom
(user-provided) hook and the default hook, which is always present. Hook
functions nest: each one must always return a plain object, and chaining happens
as a result of each function calling `next()`, which is a reference to the
subsequent loader’s hook.

> A hook that fails to return triggers an exception. A hook that returns without

Reviewer 2: > A hook that fails to return triggers an exception.

What does this mean?  I can think of several interpretations: the hook throws a sync exception, it returns a rejected promise, it returns a promise that does not resolve after a certain timeout, it returns a non-promise, or it returns a promise that resolves to a non-object value?

Author: it literally throws and the process exists with a non-zero status

Reviewer 2: Thanks, sorry, I asked my question incorrectly.  I meant, what does it mean for a hook to fail to return?  Like, what does a hook need to do to "fail to return?"

Author: The return value is not as described for each hook below (an object with the necessary properties).

Author: Is this any clearer?

```suggestion
A hook neglecting to return triggers an exception. A hook that returns without
```

Reviewer 3: I don't think it's possible to "fail to return", I initially though you meant "awaits a never settling promise". I think we could say:
```suggestion
A hook that returns a nullish value triggers an exception. A hook that returns without
```

or even (since returning `undefined` technically fits the other case):
```suggestion
A hook that returns without
```

Reviewer 3: Even then, "properly return" is not a well defined concept x) wdyt of the following?
```suggestion
A hook that returns a value that lacks a required property triggers an exception. A hook that returns without
```

Reviewer 3: (I still think we could skip the first sentence, required properties being required seems obvious to me; it's not like we are saying that the second sentence is the only way of triggering a exception in a hook)

Author: sure, works for me. pushing

When hooks are used they apply to the entry point and all `import` calls. They
won't apply to `require` calls; those still follow [CommonJS][] rules.

Loaders follow the pattern of `--require`:

```console
node \
  --experimental-loader unpkg \
  --experimental-loader http-to-https \
  --experimental-loader cache-buster
```

These are called in the following sequence: `cache-buster` calls
`http-to-https` which calls `unpkg`.

### Hooks

Hooks are part of a chain, even if that chain consists of only one custom
(user-provided) hook and the default hook, which is always present. Hook
functions nest: each one must always return a plain object, and chaining happens
as a result of each function calling `next<hookName>()`, which is a reference
to the subsequent loader’s hook.

A hook that fails to return triggers an exception. A hook that returns without
calling `next<hookName>()` and without returning `shortCircuit: true` also
triggers an exception. These errors are to help prevent unintentional breaks in
the chain.

#### `resolve(specifier, context, nextResolve)`

<!-- YAML
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/42623

>     description: Add support for chaining resolve hooks. This introduces a
>       breaking change, wherein a hook that does not call `nextResolve()` MUST
>       include a `shortCircuit` flag in its return.

Reviewer 3: Two things:
- I don't think the breaking change part is necessary; it's a good thing to insist on it in changelogs, but I think it's implied in the history section.
- I think it's not obvious what ``include a `shortCircuit` flag in its return`` means.

```suggestion
    description: Add support for chaining resolve hooks. Each hook must either
                 call `nextResolve()` or have a `shortCircuit` property set to
                 `true` in its returned object.
```

Author: Why does lint-md disallow `:` in the description? 😞

Reviewer 3: I think that's a YAML problem (or at least, the flavor of YAML `lint-md` supports 😬). You can work around this by using `>` (or is it `|`? or maybe yet another char, not sure.)
```suggestion
    description: >
      Add support for chaining resolve hooks. This introduces a
      breaking change, wherein a hook that does not call `nextResolve()` MUST
      include a `shortCircuit` flag in its return.
```

changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/42623
    description: Add support for chaining resolve hooks. Each hook that does

>       not call `nextResolve()` MUST include a `shortCircuit` property set to

Reviewer 3: I think it's a weird way of presenting the options, why not go the `must either` route? Why is must in capital letters?

Author: since there's no markdown to emphasise the text

Reviewer 3: I don't think the emphasis is needed.

* `specifier` {string}
* `context` {Object}
  * `conditions` {string\[]} Export conditions of the relevant `package.json`
  * `importAssertions` {Object}
  * `parentURL` {string|undefined} The module importing this one, or null if
    this is the Node.js entry point
* `next` {Function} The subsequent `resolve` hook in the chain, or the Node.js
  default `resolve` hook after the last user-supplied `resolve` hook
  * `specifier` {string}
  * `context` {Object}
* Returns: {Object}
  * `format` {string|null|undefined} A hint to the load hook (it might be

>     ignored)

Author: yes, we currently accept nullish

    `'builtin' | 'commonjs' | 'json' | 'module' | 'wasm'`
  * `shortCircuit` {undefinded|true} A signal that this hook intends to
    terminate the chain of `resolve` hooks
  * `url` {string} The absolute URL to which this input resolves

The `resolve` hook chain is responsible for resolving file URL for a given
module specifier and parent URL, and optionally its format (such as `'module'`)
as a hint to the `load` hook. If a format is specified, the `load` hook is
ultimately responsible for providing the final `format` value (and it is free to
ignore the hint provided by `resolve`); if `resolve` provides a `format`, a
custom `load` hook is required even if only to pass the value to the Node.js
default `load` hook.

The module specifier is the string in an `import` statement or

> `import()` expression (colloquially called "import path"):

Author: I'm fine to delete it (and was thinking maybe I shouldn't add it so as to not encourage wrong terminology).

Before I joined Node.js, this was exclusively term I'd heard used, and I'm sure I've seen it in StackOverflow questions.

```mjs
import foo from 'module specifier';

const bar = await import('module specifier');
```

> 

Reviewer 4: Yes. But also it’s repeating what’s already in the text, that the specifier is the string in an `import` statement/expression. This is basic JavaScript syntax that we link to MDN for, I don’t think we need to get into much detail explaining it in our docs.

      ...context,
      conditions: [...context.conditions, 'another-condition'],
    });
  }

  // Defer to the next hook in the chain, which would be the
  // Node.js default resolve if this is the last user-specified loader.
  return next(specifier, context);
}
```

#### `load(url, context, next)`

> The loaders API is being redesigned. This hook may disappear or its
> signature may change. Do not rely on the API described below.

> In a previous version of this API, this was split across 3 separate, now
> deprecated, hooks (`getFormat`, `getSource`, and `transformSource`).

* `url` {string} The URL returned by the `resolve` chain
* `context` {Object}
  * `conditions` {string\[]} Export conditions of the relevant `package.json`
  * `format` {string|null|undefined} The format optionally supplied by the
    `resolve` hook chain
  * `importAssertions` {Object}
* `next` {Function} The subsequent `load` hook in the chain, or the Node.js
  default `load` hook after the last user-supplied `load` hook
  * `specifier` {string}
  * `context` {Object}
* Returns: {Object}
  * `format` {string}
  * `shortCircuit` {undefined|boolean} A signal that this hook intends to

>     terminate the chain of `resolve` hooks

Reviewer 3: ```suggestion
    terminate the chain of `resolve` hooks. **Default:** `false`
```

`context.conditions` array originally passed into the `resolve` hook.

```js

> export async function resolve(specifier, context, next) {

Author: Hmm, I thought of that (I think I might have done initially?), but I think opted for `next` for consistency:

```js
function resolve(…, context, next) {}
function load   (…, context, next) {}
```

I don't have a preference

> export async function resolve(specifier, context, next) {

Reviewer 5: Since most people working with loader code will be working by example, having the code be as self-documenting as possible would be a benefit I think.

> export async function resolve(specifier, context, next) {

Reviewer 4: This would be similar to the current `defaultResolve`, `defaultLoad` etc. 👍

  const { parentURL = null } = context;

  if (Math.random() > 0.5) { // Some condition.
    // For some or all specifiers, do some custom logic for resolving.
    // Always return an object of the form {url: <string>}.
    return {
      shortCircuit: true,
      url: parentURL ?
        new URL(specifier, parentURL).href :
        new URL(specifier).href,
    };
  }

  if (Math.random() < 0.5) { // Another condition.
    // When calling `defaultResolve`, the arguments can be modified. In this
    // case it's adding another value for matching conditional exports.
    return nextResolve(specifier, {
      ...context,
      conditions: [...context.conditions, 'another-condition'],
    });
  }

  // Defer to the next hook in the chain, which would be the
  // Node.js default resolve if this is the last user-specified loader.
  return nextResolve(specifier, context);
}
```

#### `load(url, context, nextLoad)`

> 

Author: There are a few places that mention this, and they all use `shortCircuit: true`. Mind if we use that here too? (I think it's more straightforward, and concise—like "copy+paste this, and you're good")

> 

Reviewer 3: Sure, my point was we need a history section that list the changes, no strong opinion on how to word it best, I'm happy to leave that out to you :)

> 

Author: Actually, there is already an entry for this under the `## Loaders` heading. Since it applies to both and was introduced at the same time, would it not be appropriate to consolidate it there?

`context.conditions` array originally passed into the `resolve` hook.

```js
export async function resolve(specifier, context, nextResolve) {
  const { parentURL = null } = context;

  if (Math.random() > 0.5) { // Some condition.
    // For some or all specifiers, do some custom logic for resolving.
    // Always return an object of the form {url: <string>}.
    return {
      shortCircuit: true,
      url: parentURL ?
        new URL(specifier, parentURL).href :
        new URL(specifier).href,
    };
  }

  if (Math.random() < 0.5) { // Another condition.
    // When calling `defaultResolve`, the arguments can be modified. In this
    // case it's adding another value for matching conditional exports.
    return nextResolve(specifier, {
      ...context,
      conditions: [...context.conditions, 'another-condition'],
    });
  }

  // Defer to the next hook in the chain, which would be the
  // Node.js default resolve if this is the last user-specified loader.
  return nextResolve(specifier, context);
}
```

#### `load(url, context, nextLoad)`

<!-- YAML
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/42623

>     description: Add support for chaining load hooks.  Each hook must either

Reviewer 3: nit
```suggestion
    description: Add support for chaining load hooks. Each hook must either
```

Author: 😤

> #### HTTPS loader

Reviewer 4: This loader and the CoffeeScript one are in https://github.com/nodejs/loaders-test. Do those loaders work under this branch (maybe with `shortCircuit: true` added)?

Author: The https loader does not due to #42721. I'll confirm the CoffeeScript one still does.

Author: good news: nodejs/loaders-test#6

// CoffeeScript files end in .coffee, .litcoffee or .coffee.md.
const extensionsRegex = /\.coffee$|\.litcoffee$|\.coffee\.md$/;

export async function resolve(specifier, context, next) {
  // When it's not CoffeeScript, defer to another resolve hook.
  if (extensionsRegex.test(specifier)) return next(specifier, context);

> 

Reviewer 4: Because most hooks are written with the call to `next` as the fallback/default condition. Like if it were a `switch`, the call to `next` goes at the end in `default:`.

When there’s only one condition being evaluated, it feels arbitrary; but imagine of there were two or more conditions here, like “if the specifier ends in a CoffeeScript extension” _or_ “if the parentUrl ends in a CoffeeScript extension” _or_ “if the `format` returned by a preceding hook is `coffee`“ etc. You wouldn’t want to put all of these in the condition at top in the negative to call `next` on, and then branch on each of these conditions _again_ in the body of the hook. It’s less duplicative to check for each one in turn and then call `next` at the end if no condition evaluates as true, like the structure of a `switch`.

> 

Author: Not trying to play contrarian, but those don't sound like abort conditions then? If you can simply abort early, do; if you can't, don't.

My thinking for these examples is to highlight how and what needs to happen, in the most grokkable way: when your hook does not handle something, you need to call `next()`. I think putting that at the bottom makes it easily overlooked: when reading an example, people are looking to see how the author achieved the thing, and they stop reading when they do.

> 

Reviewer 4: At the very least, rewriting the example to change its style seems beyond the scope of adding chaining. Can this rewrite be put into its own follow-up PR, and the examples in this PR be rewritten as minimally as necessary to support chaining? And we can debate what makes the most sense to most people in a PR devoted to that.

  const { parentURL = baseURL } = context;

  // Node.js normally errors on unknown file extensions, so return a URL for
  // specifiers ending in the CoffeeScript file extensions.
  return {
    shortCircuit: true,
    url: new URL(specifier, parentURL).href
  };
}

export async function load(url, context, next) {
  // When it's not CoffeeScript, defer to another load hook.

>   if (!extensionsRegex.test(url)) return next(url, context);

Author: ahh, the `coffeeCompile` line changed. updated.

will update the negative/positive condition thing based on the outcome of other discussion (keeping in one place).

          // This example assumes all network-provided JavaScript is ES module
          // code.
          format: 'module',
          shortCircuit: true,
          source: data,
        }));
      }).on('error', (err) => reject(err));
    });
  }

  // Let Node.js handle all other URLs.
  return nextLoad(url, context);
}
```

> ```mjs

Author: Most of the other code blocks here use `mjs`. Shall I switch them to `js` in a later tidy PR?

> ```mjs

Reviewer 3: If the inconsistency bothers you, sure go ahead. It's a non-blocking suggestion anyway, I think in general we should try to keep the changes minimal, but I don't mind this landing with or without that line change.

prints the current version of CoffeeScript per the module at the URL in
`main.mjs`.

It is also possible to facilitate via custom loaders. While this may seem like a
significant improvement to Node.js core functionality, there are substantial
downsides to actually using this loader: performance is much slower than loading

> files from disk, there is no caching, and there is no security.

Author: What the heck. git had a stroke and added this twice 😞
